shared-bindings/alarm/__init__.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Dan Halbert for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+#include "py/reload.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/alarm/__init__.h"
+#include "shared-bindings/alarm/pin/PinAlarm.h"
+#include "shared-bindings/alarm/time/TimeAlarm.h"
+#include "shared-bindings/supervisor/Runtime.h"
+#include "shared-bindings/time/__init__.h"
+#include "supervisor/shared/autoreload.h"
+#include "supervisor/shared/workflow.h"
+
+// Wait this long imediately after startup to see if we are connected to USB.
+#define CIRCUITPY_USB_CONNECTED_SLEEP_DELAY 5
+
+//| """Alarms and sleep
+//|
+//| Provides alarms that trigger based on time intervals or on external events, such as pin
+//| changes.
+//| The program can simply wait for these alarms, or go into a sleep state and
+//| and be awoken when they trigger.
+//|
+//| There are two supported levels of sleep: light sleep and deep sleep.
+//|
+//| Light sleep leaves the CPU and RAM powered so the program can resume after sleeping.
+//|
+//| *However, note that on some platforms, light sleep will shut down some communications, including
+//| WiFi and/or Bluetooth.*
```

Comment:
I'm not sure light sleep should do any shutdown. We should leave it up to the python code to manage it because it'll have to reconnect after light sleep either way.

Let's just have light sleep wait if wifi is active. We make no guarantees how light the sleep is.

Reply:
ESP32-S2 light sleep is pretty deep in terms of power saving, but allows the program to resume from where it left off. That's why I included it here. But I agree the side effects are significant. I should document what you need to do to really do a light sleep. It looks like if `common_hal_wifi_radio_get_enabled(wifi.radio)` is what to check about whether wifi is active. Does that sound right to you?

Reply:
Another possibility is to not have separate `wait_until_alarms()` and `sleep_until_alarms()`. Instead, we just have a isngle routine (we have to pick one of the names). If light sleep is possible because wifi is off, etc., then go ahead and do it. Otherwise just wait. There could be a property `.light_sleep_possible` that would tell you whether you were going to sleep or not, though I'm not sure what the program would do with that if it didn't like the answer.

Reply:
I was wondering if we should even distinguish between light and deep sleep, because the routine names show what happens (you exit or you just wait). So whichever routine you call, it tries to sleep as deeply as possible, given the constraints of the semanatics for that routine.

Reply:
Naming possibilities:
- Combine `wait_until_alarms()` and `sleep_until_alarms()` into one name or the other, or into `try_to_sleep_until_alarms()` (or `attempt_sleep_until_alarms()`).
- Drop the `deep` in `exit_and_deep_sleep_until_alarms()`, and/or add `try_to`

The q is whether to try to point out that these routines might not really sleep, or whether to use the word `sleep` as `time.sleep()` does, which may not really sleep.

Reply:
After a conversation with @tannewt, we dropped `wait_until_alarms()` and now have two calls:
- `light_sleep_until_alarms()`
- `exit_and_deep_sleep_until_alarms()`

The first will light-sleep to the extent possible, but will not shut down wifi, etc. The addition of `light` is meant to differentiate from `deep`.